napari/utils/_tests/test_progress.py

```diff
@@ -0,0 +1,101 @@
+import numpy as np
+import pytest
+
+try:
+    import qtpy  # noqa
+except ImportError:
+    pytest.skip('Cannot test progress without qtpy.', allow_module_level=True)
+except RuntimeError:
+    pytest.skip(
+        'Cannot test progress without Qt bindings.', allow_module_level=True
+    )
+from ..progress import ProgressBar, progress
+
+
+def activity_dock_children(viewer):
+    return viewer.window.qt_viewer.activityDock.children()[4].children()
+
+
+def test_progress_with_iterable(make_napari_viewer):
+    viewer = make_napari_viewer()
+    r = range(100)
+    pbr = progress(r)
+
+    assert any(
+        isinstance(wdg, ProgressBar) for wdg in activity_dock_children(viewer)
+    )
+    assert pbr.iterable is r
+    assert pbr.n == 0
+    assert pbr._pbar.pbar.maximum() == pbr.total == 100
+
+
+def test_progress_with_ndarray(make_napari_viewer):
+    viewer = make_napari_viewer()
+    iter = np.random.random((100, 100))
+    pbr = progress(iter)
+
+    assert any(
+        isinstance(wdg, ProgressBar) for wdg in activity_dock_children(viewer)
+    )
+    assert pbr.iterable is iter
+    assert pbr.n == 0
+    assert pbr._pbar.pbar.maximum() == pbr.total
+
+
+def test_progress_with_total(make_napari_viewer):
+    viewer = make_napari_viewer()
+    pbr = progress(total=5)
+
+    assert any(
+        isinstance(wdg, ProgressBar) for wdg in activity_dock_children(viewer)
+    )
+    assert pbr.n == 0
+    assert pbr._pbar.pbar.maximum() == pbr.total == 5
+
+    pbr.update(1)
+    assert pbr.n == 1
+
+    pbr.close()
+
+
+def test_progress_with_context(make_napari_viewer):
+    viewer = make_napari_viewer()
+
+    with progress(range(100)) as pbr:
+        assert any(
+            isinstance(wdg, ProgressBar)
+            for wdg in activity_dock_children(viewer)
+        )
+        assert pbr.n == 0
+        assert pbr._pbar.pbar.maximum() == pbr.total == 100
+
+
+def test_progress_update(make_napari_viewer):
+    make_napari_viewer()
+
+    pbr = progress(total=10)
+
+    assert pbr.n == 0
+    assert pbr._pbar.pbar.value() == 0
+
+    pbr.update(1)
+    pbr.refresh()  # not sure why this has to be called manually here
+
+    assert pbr.n == 1
+    assert pbr._pbar.pbar.value() == 1
+
+    pbr.update(2)
+    pbr.refresh()
+
+    assert pbr.n == 3
+    assert pbr._pbar.pbar.value() == 3
+
+
+def test_progress_set_description(make_napari_viewer):
+    make_napari_viewer()
+
+    pbr = progress(total=5)
+    pbr.set_description("Test")
+
+    assert pbr.desc == "Test: "
+    assert pbr._pbar.description_label.text() == "Test: "
```

Comment:
somewhere in here you should also test that the `progress()` wrapper can simply wrap an iterable and that it doesn't modify the result.  for instance:

```python
assert list(progress(range(10))) == list(range(10))
```

in this case, that test would reveal the AttributeError I mentioned above:

```pytb
~/Dropbox (HMS)/Python/forks/napari/napari/utils/progress.py in display(self, msg, pos)
    128     def display(self, msg: str = None, pos: int = None) -> None:
    129         """Update the display."""
--> 130         if not self.viewer:
    131             return super().display(msg=msg, pos=pos)
    132

AttributeError: 'progress' object has no attribute 'viewer'
```